libafl/src/stages/power.rs

```diff
@@ -57,54 +80,296 @@ where
+    }
+
+    #[inline]
+    #[allow(
+        clippy::cast_precision_loss,
+        clippy::too_many_lines,
+        clippy::cast_sign_loss
+    )]
+    fn calculate_score(
+        &self,
+        testcase: &mut Testcase<I>,
+        statsdata: &PowerScheduleStats,
+        fuzz_mu: f64,
+    ) -> Result<usize, Error> {
+        let mut perf_score = 100.0;
+        let q_exec_us = testcase
+            .exec_time()
+            .ok_or_else(|| Error::KeyNotFound("exec_time not set".to_string()))?
+            .as_nanos() as f64;
+
+        let avg_exec_us = statsdata.exec_time().as_nanos() as f64 / statsdata.cycles() as f64;
+        let avg_bitmap_size = statsdata.bitmap_size() / statsdata.bitmap_entries();
+
+        let metadata = testcase
+            .metadata_mut()
+            .get_mut::<PowerScheduleTestData>()
+            .ok_or_else(|| Error::KeyNotFound("PowerScheduleTestData not found".to_string()))?;
+
+        if q_exec_us * 0.1 > avg_exec_us {
+            perf_score = 10.0;
+        } else if q_exec_us * 0.2 > avg_exec_us {
+            perf_score = 25.0;
+        } else if q_exec_us * 0.5 > avg_exec_us {
+            perf_score = 50.0;
+        } else if q_exec_us * 0.75 > avg_exec_us {
+            perf_score = 75.0;
+        } else if q_exec_us * 4.0 < avg_exec_us {
+            perf_score = 300.0;
+        } else if q_exec_us * 3.0 < avg_exec_us {
+            perf_score = 200.0;
+        } else if q_exec_us * 2.0 < avg_exec_us {
+            perf_score = 150.0;
+        }
+
+        let q_bitmap_size = metadata.bitmap_size() as f64;
+        if q_bitmap_size * 0.3 > avg_bitmap_size as f64 {
+            perf_score *= 3.0;
+        } else if q_bitmap_size * 0.5 > avg_bitmap_size as f64 {
+            perf_score *= 2.0;
+        } else if q_bitmap_size * 0.75 > avg_bitmap_size as f64 {
+            perf_score *= 1.5;
+        } else if q_bitmap_size * 3.0 < avg_bitmap_size as f64 {
+            perf_score *= 0.25;
+        } else if q_bitmap_size * 2.0 < avg_bitmap_size as f64 {
+            perf_score *= 0.5;
+        } else if q_bitmap_size * 1.5 < avg_bitmap_size as f64 {
+            perf_score *= 0.75;
+        }
+
+        if metadata.handicap() >= 4 {
+            perf_score *= 4.0;
+            metadata.set_handicap(metadata.handicap() - 4);
+        } else if metadata.handicap() > 0 {
+            perf_score *= 2.0;
+            metadata.set_handicap(metadata.handicap() - 1);
+        }
+
+        if metadata.depth() >= 4 && metadata.depth() < 8 {
+            perf_score *= 2.0;
+        } else if metadata.depth() >= 8 && metadata.depth() < 14 {
+            perf_score *= 3.0;
+        } else if metadata.depth() >= 14 && metadata.depth() < 25 {
+            perf_score *= 4.0;
+        } else if metadata.depth() >= 25 {
+            perf_score *= 5.0;
+        }
+
+        let mut factor: f64 = 1.0;
+
+        // TODO: currently we don't have any favored inputs, if that's introduced we need to modify code here
```

Comment:
Sounds like something we should, somehow, add conceptual support for, right?

Reply:
I will work on `favored`.

Reply:
added code for using `favored` from aflfast side.
the actual code for determining if a corpus entry is favored should be another pr?

Reply:
No we have it, https://github.com/AFLplusplus/LibAFL/blob/main/libafl/src/corpus/minimizer.rs#L22

Reply:
Just check for this meta when using a MinimizerCorpusScheduler like in https://github.com/AFLplusplus/LibAFL/blob/main/libafl/src/corpus/minimizer.rs#L136